src/Microsoft.AspNetCore.Server.Kestrel/KestrelServer.cs

```diff
@@ -108,8 +114,45 @@ public void Start<TContext>(IHttpApplication<TContext> application)
                     else
                     {
                         atLeastOneListener = true;
-                        _disposables.Push(engine.CreateServer(
-                            parsedAddress));
+
+                        if (parsedAddress.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
+                        {
+                            var exceptions = new List<Exception>();
+                            var originalHost = parsedAddress.Host;
+
+                            var loopbackAddresses = new[]
+                            {
+                                new { Address = IPAddress.Loopback.ToString(), Family = "IPv4" },
+                                new { Address = IPAddress.IPv6Loopback.ToString(), Family = "IPv6" }
+                            };
+
+                            foreach (var loopbackAddress in loopbackAddresses)
+                            {
+                                try
+                                {
+                                    parsedAddress.Host = loopbackAddress.Address;
+                                    _disposables.Push(engine.CreateServer(
+                                        parsedAddress));
+                                }
+                                catch (UvException ex) when (ex.StatusCode != Constants.EADDRINUSE)
+                                {
+                                    _logger.LogWarning(0, ex, $"Unable to bind {address} to loopback {loopbackAddress.Family} address.");
+                                    exceptions.Add(ex);
+                                }
+                            }
+
+                            if (exceptions.Count == 2)
+                            {
+                                throw new InvalidOperationException("Binding to localhost failed to bind to any loopback interface.");
+                            }
+
+                            parsedAddress.Host = originalHost;
```

Comment:
I'm using the same object to get `Port` updated in case it was 0. Otherwise binding to `localhost:0` keeps the port as 0 in the end.